### DISCUSSION TITLE
Wall profiling performance impact

### DISCUSSION THREAD

**Discussion by michaldo** [async-profiler/async-profiler#964] [General]
Hi,
For long time I was sure async profiler performance impact is low. But now I measure it and figures are different.
There is noticeable differenence between _wall_ profiling and _alloc_ profiling
Command:
left: start,event=wall,loop=5m,file=/app/profiler-output/%t-app.jfr
right:  start,event=alloc,loop=5m,file=/app/profiler-output/%t-app.jfr
Output
metrics: system_cpu_usage, proces_cpu_usage
Processor: m6i.xlarge
JVM: OpenJDK Runtime Environment Zulu21.28+85-CA (build 21+35)
Async profiler started with https://github.com/jvm-profiling-tools/ap-loader version 3.0-9

wall: ~17%
alloc: ~7%

![image](https://github.com/user-attachments/assets/36760a90-41bd-4791-beb9-21d651b896e8)

Context: Spring Boot application. Same result noticed on other applications

Question: how big is impact of profiling in wall mode?

**Comment by apangin** [comment]
The overhead highly depends on the particular application and profiler configuration.
CPU profiling with 100ms interval and allocation profiling with 2mb interval typically has no significant impact on most applications.
However, wall clock profiling with the default interval may indeed cause high overhead if an application has many (hundreds) threads. That's because wall clock profiler always samples all threads including idle ones.

The problem is understood and I have an idea how to improve wall clock profiler - the clue is to skip threads from sampling if they have not been executed on CPU since last cycle. I will implement the optimization in one of the next releases. Meanwhile, the workaround is to simply increase profiling interval (the overhead is inversely proportional to the sampling interval).

**Reply by michaldo** [reply]
My application has 245 threads

**Comment by hridayahuja44** [comment]
What is the ideal interval for wall clock profiling if we are enabling CPU profiling and wall clock profiling where we have fixated the interval for CPU profiling to be 5ms, our application also uses spring boot. I am also concerned about the overhead so wanted to know the same. Thank you so much...

**Reply by apangin** [reply]
There is no universal optimal value for the interval - it depends on an application. I'd start with a safe value of 1000ms and then gradually decrease it while there is no impact on the application performance.

A good news is that I've implemented the optimization mentioned above, will commit it shortly. Stay tuned.
